lib/usd/translators/shading/usdMaterialReader.cpp

```diff
@@ -0,0 +1,136 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "usdMaterialReader.h"
+
+#include <mayaUsd/fileio/shaderReader.h>
+#include <mayaUsd/fileio/translators/translatorUtil.h>
+#include <mayaUsd/fileio/utils/readUtil.h>
+#include <mayaUsd/utils/util.h>
+
+#include <pxr/base/tf/diagnostic.h>
+#include <pxr/base/tf/staticTokens.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/base/vt/value.h>
+#include <pxr/pxr.h>
+#include <pxr/usd/sdf/path.h>
+#include <pxr/usd/sdf/types.h>
+#include <pxr/usd/sdf/valueTypeName.h>
+#include <pxr/usd/usdShade/input.h>
+#include <pxr/usd/usdShade/output.h>
+#include <pxr/usd/usdShade/shader.h>
+#include <pxr/usd/usdShade/tokens.h>
+
+#include <maya/MFnDependencyNode.h>
+#include <maya/MObject.h>
+#include <maya/MPlug.h>
+#include <maya/MStatus.h>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+// clang-format off
+TF_DEFINE_PRIVATE_TOKENS(
+    _tokens,
+
+    (outColor)
+);
+// clang-format on
+
+PxrUsdTranslators_MaterialReader::PxrUsdTranslators_MaterialReader(
+    const UsdMayaPrimReaderArgs& readArgs)
+    : UsdMayaShaderReader(readArgs)
+{
+}
+
+/* virtual */
+bool PxrUsdTranslators_MaterialReader::Read(UsdMayaPrimReaderContext* context)
+{
+    const auto&    prim = _GetArgs().GetUsdPrim();
+    UsdShadeShader shaderSchema = UsdShadeShader(prim);
+    if (!shaderSchema) {
+        return false;
+    }
+
+    const TfToken&    mayaNodeTypeName = _GetMayaNodeTypeName();
+    MStatus           status;
+    MObject           mayaObject;
+    MFnDependencyNode depFn;
+    if (!(UsdMayaTranslatorUtil::CreateShaderNode(
+              MString(prim.GetName().GetText()),
```

Comment:
Should we be using more of the prim's path in the Maya node name here to avoid collisions when the shader prims in different namespaces in USD share the same name?

Reply:
The node name will be made unique by Maya by appending or incrementing a last digit. It would be less surprising (to me) to have blueRubber become blueRubber1 than to see very long path prefixed names.

Reply:
I think the concern would be that the material/shader names could be unstable across round-trips in Maya then. Within the asset itself, this might be ok, but for anything else that references the exported asset and then imparts other opinions, you run the risk of those opinions falling off if the prim names change.

Reply:
Makes sense indeed. We would need to detect when the path prefix was already added, otherwise the material name will accumulate a path prefix at each roundtrip.
In the case where an asset is referenced multiple times with the same set of opinions, should we try to collapse its materials when possible? What would be the names? I suspect we will have to add some code to reduce the number of materials generated on import, like having only two materials generated for the two-toned tile floor of the Kitchen_set.

Reply:
I think we've talked about it before, but naively my thought was that we might be able to translate the USD namespace hierarchy into a Maya namespaced name. For example:

UsdShade prim: `/Chair/Materials/Wood/WoodGrain`
Maya shader node: `:Chair:Materials:Wood:WoodGrain`

And then as far as collapsing, and because the Kitchen_set doesn't actually have any materials in USD, I guess you're thinking about the displayColors shading mode? I guess collapsing could be a special case for that shading mode if we really wanted that, but I'd say that we might want to move away from and eventually deprecate the displayColors shading mode altogether. If there's no material in USD, it doesn't really feel like we should invent one based on the displayColors. We could just assign the Maya geometry the initialShadingGroup and turn on the geometry's `displayColors` attribute.

For all other shading modes and when there is a material prim to work with, USD somewhat enforces that the shaders connected to a Material live under it in namespace, so I would think it should be a pretty direct translation to shading nodes in Maya on import.

Reply:
I will have to check if Maya namespaces can be used like that. That would be nice. I have opened a ticket in our internal tracker to track this improvement, which will probably be in a follow-up PR.